clippy_utils/src/lib.rs

```diff
@@ -2791,16 +2791,46 @@ impl<'tcx> ExprUseCtxt<'tcx> {
                         .position(|arg| arg.hir_id == self.child_id)
                         .map_or(0, |i| i + 1),
                 ),
-                ExprKind::Field(_, name) => ExprUseNode::FieldAccess(name),
+                ExprKind::Field(_, name) => ExprUseNode::FieldAccess(use_expr.hir_id, name),
                 ExprKind::AddrOf(kind, mutbl, _) => ExprUseNode::AddrOf(kind, mutbl),
+                ExprKind::Assign(lhs, rhs, _) => {
+                    debug_assert!(lhs.hir_id == self.child_id || rhs.hir_id == self.child_id);
+                    #[allow(clippy::bool_to_int_with_if)]
+                    let idx = if lhs.hir_id == self.child_id { 0 } else { 1 };
+                    ExprUseNode::Assign(idx)
+                },
+                ExprKind::AssignOp(op, lhs, rhs) => {
+                    debug_assert!(lhs.hir_id == self.child_id || rhs.hir_id == self.child_id);
+                    #[allow(clippy::bool_to_int_with_if)]
+                    let idx = if lhs.hir_id == self.child_id { 0 } else { 1 };
+                    ExprUseNode::AssignOp(op.node, idx)
+                },
                 _ => ExprUseNode::Other,
             },
             _ => ExprUseNode::Other,
         }
     }
+
+    pub fn use_mutability(&self, cx: &LateContext<'tcx>) -> Mutability {
+        match self.use_node(cx) {
+            ExprUseNode::FieldAccess(parent, _) => {
+                let parent = cx.tcx.hir().expect_expr(parent);
+                expr_use_ctxt(cx, parent).use_mutability(cx)
+            },
+            ExprUseNode::AssignOp(_, 0) | ExprUseNode::Assign(0) => Mutability::Mut,
+            ExprUseNode::AddrOf(_, mutbl) => mutbl,
+            ExprUseNode::FnArg(_, _) | ExprUseNode::MethodArg(_, _, _) => {
+                let child_expr = cx.tcx.hir().expect_expr(self.child_id);
+                let ty = cx.typeck_results().expr_ty_adjusted(child_expr);
+                ty.ref_mutability().unwrap_or(Mutability::Not)
+            },
+            _ => Mutability::Not,
+        }
+    }
```

Comment:
Using `expr_use_ctxt` here is wrong since it allows moves to occur. Just use [`parent_iter`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/map/struct.Map.html#method.parent_iter).

You'll need to handle mutable borrow or deref adjustments, mutable borrows, mutable derefs, mutable indexing, and assignments. For indexing and derefs you can't tell at the expression so you have to keep walking up the tree. For adjustments, don't go by the resulting type, but walk the adjustments themselves. Mutable auto-deref doesn't always result in a reference after all adjustments take place.

Reply:
> Using `expr_use_ctxt` here is wrong since it allows moves to occur. Just use [`parent_iter`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/map/struct.Map.html#method.parent_iter).
> 
> You'll need to handle mutable borrow or deref adjustments, mutable borrows, mutable derefs, mutable indexing, and assignments. For indexing and derefs you can't tell at the expression so you have to keep walking up the tree. For adjustments, don't go by the resulting type, but walk the adjustments themselves.

Oh, thanks for pointing those out, I'll check it out.

> Mutable auto-deref doesn't always result in a reference after all adjustments take place.

Would you mind explaining it further? Under what circumstances would it not work? So that I can write a test case to make sure of that.

Reply:
Field accesses through a pointer don't. e.g. `x.y` where `x` is a reference will have have a deref adjustment, but not a borrow adjustment.